docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,430 @@
+kind: BeforeClusterDeleteResponse
+status: Success
+message: "error message if status == Failure"
+retryAfterSeconds: 10
+```
+
+For additional details, refer to the [Draft OpenAPI spec](https://editor.swagger.io/?url=https://gist.githubusercontent.com/ykakarap/6a58d976df90928cdb31768125d73613/raw/8ce6175060039c2fb7b3ea9a091d712228f56852/lifecycle-hooks-openapi-2.yaml).
+
+
+###  Runtime Extensions developer guide
+
+All guidelines defined in the [Runtime SDK](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md) apply to the implementation of Runtime Extensions of the hooks defined in this proposal. 
+
+TL;DR; Runtime Extensions are components that should be designed, written and deployed with great caution given that they can affect the proper functioning of the Cluster API runtime. A poorly implemented Runtime Extension could potentially block lifecycle transitions from happening.
+
+Following recommendations are especially relevant:
+
+* [Blocking and non Blocking](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#blocking-hooks)
+* [Error management](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#error-management)
+* [Avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies)
+
+
+###  Security Model
+
+For the general Runtime Extension security model please refer to the [developer guide in the Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/75b39db545ae439f4f6203b5e07496d3b0a6aa75/docs/proposals/20220221-runtime-SDK.md#security-model).
+
+###  Risks and Mitigations
+
+####  Runtime Extension blocks Cluster lifecycle indefinitely
+
+Cluster lifecycle can be blocked indefinitely when a Runtime Extension either blocks or fails indefinitely. Mitigation:
+
+* A Runtime Extension should be extensively unit and e2e tested to ensure it behaves as expected.
+
+
+## Alternatives
+
+Alternatives to Runtime Hooks for comprehensive add-on management in Cluster API include:
+
+### External components watching CAPI resources without hooks
+
+This is the current pattern used by Cluster Resource Set. The implementers can only react to changes on CAPI resources, e.g. a Cluster being created, but they have no control over the cluster lifecycle.
+
+This and similar solutions based on scripting or git ops approaches are considered inadequate for comprehensive, integrated add-on management as they have limited insight into the state of the Cluster and can not easily influence reconciliation based on add-on state.
+
+More details about this in the [Cluster Addon Proposal](https://docs.google.com/document/d/1TdbfXC2_Hhg0mH7-7hXcT1Gg8h6oXKrKbnJqbpFFvjw/edit).
```

Comment:
it's unclear to me from reading this section if the Cluster Addon Proposal is considered an alternative or just some place where more details on the current pattern are provided.

Would be great if the proposal could make it clear what is the relationship between this work and Cluster Addon work since there seems to be a lot of overlap.

Reply:
My impression is that in the current proposal we only define the Runtime Hooks which are necessary to build Addon management while the linked proposal builds on top of those hooks to actually implement addon management.

Reply:
We should improve this sentence into 

```suggestion
More details about why watching Cluster API resources without hooks is not considered a valid alternative can be found in the [Cluster Addon Proposal](https://docs.google.com/document/d/1TdbfXC2_Hhg0mH7-7hXcT1Gg8h6oXKrKbnJqbpFFvjw/edit).
```

(The Cluster Addon Proposal is just someplace where more details on the current pattern are provided; also my expectation is that @Jont828 and @jackfrancis will start using the hooks defined in this proposal as soon as possible to better orchestrate the addon lifecycle; I still have to add details about how/what problems we can solve in that document).